include/swift/ABI/Task.h

```diff
@@ -148,19 +150,93 @@ class AsyncTask : public HeapObject, public Job {
+      IsEmptyTerminal   = 0b00,
+      /// The storage pointer points at the next TaskLocalChainItem in this task.
+      IsNextItemPointer = 0b01,
+      /// The storage pointer points at a parent AsyncTask,
+      /// in which we should continue the lookup.
+      IsParentPointer   = 0b11
+  };
+
+  struct TaskLocalChainItem {
+      /// Mask used for the low status bits in a task local chain item.
+      static const uintptr_t statusMask = 0x03;
+
+      /// Metadata* to the Key of the task local value
+      uintptr_t keyType;
+      /// OpaqueValue* to the Value of the task local value
+      HeapObject *storage;
+
+      // TaskLocalChainItem*
+//      uintptr_t next;
+  };
+
+  /// Reserved for a chain of task local values,
+  /// if empty, the search should continue in this task's parent (if it has one).
+  TaskLocalChainItem *TaskLocal;
+
   AsyncTask(const HeapMetadata *metadata, JobFlags flags,
             TaskContinuationFunction *run,
             AsyncContext *initialContext)
     : HeapObject(metadata), Job(flags, run),
       ResumeContext(initialContext),
-      Status(ActiveTaskStatus()) {
+      Status(ActiveTaskStatus()),
+      TaskLocal(nullptr) {
     assert(flags.isAsyncTask());
   }
 
   void run(ExecutorRef currentExecutor) {
     ResumeTask(this, currentExecutor, ResumeContext);
   }
 
+  HeapObject* localGet(const Metadata *keyType) {
+    assert(keyType && "Task.Local key must not be null.");
+
+    if (!TaskLocal) {
+      return nullptr;
+    }
+
+    fprintf(stderr, "error: %s [%s:%d] TaskLocal=%d keyType=%d, lookup keyType=%d\n",
+            __FUNCTION__, __FILE_NAME__, __LINE__,
+            TaskLocal, TaskLocal->keyType, reinterpret_cast<uintptr_t>(keyType));
+    // TODO: can the contained list/dictionary for the key
+    if (TaskLocal->keyType == reinterpret_cast<uintptr_t>(keyType)) {
+      fprintf(stderr, "error: %s [%s:%d] return value TaskLocal=%d value=%d value.retainCount=%d\n",
+              __FUNCTION__, __FILE_NAME__, __LINE__,
+              TaskLocal, TaskLocal->storage, swift_retainCount(TaskLocal->storage));
+      return TaskLocal->storage;
+    }
+
+    if (hasChildFragment()) {
+      return childFragment()->getParent()->localGet(keyType);
+    } else {
+      // none of our parent tasks contained the value either, return null
+      return nullptr;
+    }
+  }
+
+  void localSet(const Metadata *keyType, HeapObject *value) {
+    assert(value && "Task local value must not be nil");
+
+    // TODO: proper implementation, with a linked list
+//    assert(!TaskLocal && "TaskLocal was already initialized");
+    TaskLocal = new TaskLocalChainItem;
+    TaskLocal->keyType = reinterpret_cast<uintptr_t>(keyType);
+    TaskLocal->storage = value;
+
+
+    fprintf(stderr, "error: %s [%s:%d] TaskLocal=%d [%d] value.retainCount=%d\n", __FUNCTION__, __FILE_NAME__, __LINE__,
+            TaskLocal, TaskLocal->storage,    swift_retainCount(TaskLocal->storage));
+    swift_retain(TaskLocal->storage);
```

Comment:
so we retain that heap object... it correctly shows up with retain count = 2 here

Reply:
Got it, since the `__owned` transfers ownership so it's +1 retained I guess